### ISSUE TITLE
Add entire map content as a field in the LogstashLayout

### ISSUE THREAD

**Issue by ercpe** [vy/log4j2-logstash-layout#45]
We're logging our messages via `MapMessage` through a `LogstashLayout` to ELK.

I'm trying to add the entire map content of the `MapMessage` to the layout but can't figure out how. I know that i can look up individual keys via 

```"event": "${map:event}",```

but I haven't found a way to add the entire map.
I want to do something like:

```
{
    "vars": "${map}"
}
```

and get sth. like 

```
{
"vars": {
  "key1": "value1",
  "key2": "value2"
}
```
The message keys are more or less dynamic so i can't add them all as individual lookups.

From what I can see it's not possible to register a custom `Resolver`, isn't it?

**Comment by vy** [comment]
I guess what you are looking for is `json:message:json` directive, that is, `{"vars": "${json:message:json}"}`. Does that solve your issue?

**Comment by ercpe** [comment]
Partially:

```
        Map<String, Object> m = new HashMap<>();
        m.put("event", "some_event");
        m.put("foo", "bar");

        Map<String, Object> foooMap = new HashMap<>();
        foooMap.put("a", 1);
        foooMap.put("b", true);
        foooMap.put("c", "ccc");

        m.put("vars", foooMap);

        LogManager.getLogger().info(new MapMessage<>(m));
```
and 
```
"test": "${json:message:json}"
```
gives me

```
  "test" : {
    "event" : "some_event",
    "foo" : "bar",
    "vars" : "{a=1, b=true, c=ccc}"
  }
```
It would be nice to have nested objects as proper json objects, too.

**Comment by vy** [comment]
@ercpe, the problem turned out to be this: `MapMessage` (extending `MultiformatMessage`) by default implements its own formatter, which is, if I am not mistaken, used by `{Yaml,Json,Xml}Layout` as well. (See `MapMessage#asJson()`.) `LogstashLayout` falls back to these provided defaults whenever it can. But... Yes, you guessed it right: `MapMessage#asJson()` is a cheap hack over `entry.getValue().toString()`. I have already raised this issue in [LOG4J2-2703](https://issues.apache.org/jira/browse/LOG4J2-2703).

So what is the right approach to address our problem?

1. Submit a Log4j 2 fix to handle `Map`s properly in `MapMessage`? (Then do this for every potentially broken type; e.g., `Collection`s, arrays, etc.)

2. Work around the problem by having a specialized handler for `MapMessage` in `MessageResolver` of `LogstashLayout` and let the Jackson do rest of the magic? What about custom `MapMessage` implementations that want their own `MapMessage#asJson()` get used for serialization? Introduce a new configuration variable to disable this specialized handler.

I am gonna go with the 2nd option and chase the Log4j 2 ticket for the first option.

**Comment by vy** [comment]
@ercpe, would you mind confirming the fix that I have just pushed, please?

**Comment by ercpe** [comment]
I think option 1 is the cleaner approach, but i'm totally fine with the second one although i probably would've skipped the configuration option.
I'll test the fix tomorrow at work.

Thank you for your fast response!

**Comment by ercpe** [comment]
Yep, works as expected:

Layout:
```
"app_vars": "${json:message:json}"
```

```
Map<String, Object> sub2map = new HashMap<>();
sub2map.put("bat", "baz");
sub2map.put("someint", 42);

Map<String, Object> sub1map = new HashMap<>();
sub1map.put("foo", "bar");
sub1map.put("sub2", sub2map);

Map<String, Object> topLevelMap = new HashMap<>();
topLevelMap.put("top", "test");
topLevelMap.put("sub1", sub1map);

LogManager.getLogger().info(new MapMessage<>(topLevelMap));
```
gives:

```
"app_vars" : {
    "sub1" : {
      "sub2" : {
        "bat" : "baz",
        "someint" : 42
      },
      "foo" : "bar"
    },
    "top" : "test"
  }
```

although the the `message` uses the .toString() value:

```"message" : "sub1=\"{sub2={bat=baz, someint=42}, foo=bar}\" top=\"test\"",```

but i can totally live with that since i'm not really using the `message` property.

Any ETA on the next release?

**Comment by vy** [comment]
As soon as Maven Central starts to work again. :confused:

**Comment by vy** [comment]
Released v0.21 the artifacts via Sonatype and should show up in Maven Central momentarily.
